src/runtime_src/core/tools/xbmgmt2/SubCmdFlash.cpp

```diff
@@ -95,49 +95,119 @@ static void
+      sec = nullptr;
+  }
 
-    flasher.upgradeFirmware(flashType, pri.get(), sec.get());
-    std::cout << "Shell is updated succesfully\n";
-    std::cout << "Cold reboot machine to load new shell on card" << std::endl;
+  flasher.upgradeFirmware(flashType, pri.get(), sec.get());
+  std::cout << "Shell is updated succesfully\n";
+  std::cout << "Cold reboot machine to load new shell on card" << std::endl;
 }
 
+/*
+ * Update SC firmware on the board
+ */
 static void 
 update_SC(uint16_t index, const std::string& file)
 {
-    Flasher flasher(index);
-    if(!flasher.isValid())
-        return;
+  Flasher flasher(index);
+  if(!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % index));
+
+  std::shared_ptr<firmwareImage> bmc =
+    std::make_shared<firmwareImage>(file.c_str(), BMC_FIRMWARE);
+  if (bmc->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file.c_str()));
+
+  flasher.upgradeBMCFirmware(bmc.get());
+}
+
+/* 
+ * Find the correct shell to be flashed on the board
+ * Helper method for auto_flash
+ */
+static DSAInfo 
+selectShell(uint16_t idx, std::string& dsa, std::string& id)
+{
+  uint16_t candidateDSAIndex = std::numeric_limits<uint16_t>::max();
 
-    std::shared_ptr<firmwareImage> bmc =
-        std::make_shared<firmwareImage>(file.c_str(), BMC_FIRMWARE);
-    if (bmc->fail())
-        return;
+  Flasher flasher(idx);
+  if(!flasher.isValid())
+    return DSAInfo("");
+
+  std::vector<DSAInfo> installedDSA = flasher.getInstalledDSA();
+
+  // Find candidate DSA from installed DSA list.
+  if (dsa.empty()) {
+    std::cout << "Card [" << flasher.sGetDBDF() << "]: " << std::endl;
+    if (installedDSA.empty()) {
+      std::cout << "\t Status: no shell is installed" << std::endl;
+      return DSAInfo("");
+    }
+    if (installedDSA.size() > 1) {
+      std::cout << "\t Status: multiple shells are installed" << std::endl;
+      return DSAInfo("");
+    }
+    candidateDSAIndex = 0;
+  } else {
+    for (uint16_t i = 0; i < installedDSA.size(); i++) {
+      DSAInfo& idsa = installedDSA[i];
+      if (dsa != idsa.name)
+        continue;
+      if (!id.empty() && !idsa.matchId(id))
+        continue;
+      if (candidateDSAIndex != std::numeric_limits<uint16_t>::max()) {
+        std::cout << "\t Status: multiple shells are installed" << std::endl;
+        return DSAInfo("");
+      }
+      candidateDSAIndex = i;
+    }
+  }
+  if (candidateDSAIndex == std::numeric_limits<uint16_t>::max()) {
+    std::cout << "WARNING: Failed to flash Card["
+      << flasher.sGetDBDF() << "]: Specified shell is not applicable" << std::endl;
+    return DSAInfo("");
+  }
 
-    flasher.upgradeBMCFirmware(bmc.get());
+  DSAInfo& candidate = installedDSA[candidateDSAIndex];
+  bool same_dsa = false;
+  bool same_bmc = false;
+  DSAInfo currentDSA = flasher.getOnBoardDSA();
+  if (!currentDSA.name.empty()) {
+    same_dsa = (candidate.name == currentDSA.name &&

```

Comment:
A lot going on here.
To help the future user reading the code (instead of relying on order of operation), could you add a set of parenthesis around the first and second "equals" operators.  
```
same_dsa = ((candidate.name == currentDSA.name) && 
            (candidate.matchId(currentDSA)));
```